### PR TITLE
Add per source EOL settings for Dogstatsd

### PR DESCRIPTION
**Pull request by gh123man**
### What does this PR do?
Followup to this PR: https://github.com/DataDog/datadog-agent/pull/7090

This change allows per source (UDS, UDP, Named Pipe) settings for dogstatsd EOL termination for further internal testing.

### Describe your test plan

UT's, manual testing

### CODE REVIEW

**Review by ogaca-dd — COMMENTED** [review]
In this PR, the `Packet` struct contains the source type. 
It is better that `Packet` has a field `eolTermination bool` instead as:
- It would reduce the code for this feature
- it would keep `Packet` independent from the source type.

File: pkg/dogstatsd/server.go [hunk 1]
```diff
@@ -211,6 +213,21 @@ func NewServer(aggregator *aggregator.BufferedAggregator, extraTags []string) (*
...
+
+	for _, v := range config.Datadog.GetStringSlice("dogstatsd_eol_required") {
+		switch v {
+		case "udp":
+			eolTerminationUDP = true
+		case "uds":
+			eolTerminationUDS = true
+		case "named_pipe":
+			eolTerminationNamedPipe = true
+		}
```
Comment (ogaca-dd) [inline pkg/dogstatsd/server.go:228]: It would be nice to add a `default` case to display an error in case of a typo.

**Review by truthbk — COMMENTED** [review]
I like the approach, semantically very clean. 

I am thinking about @ogaca-dd comment.... while his comment might help improve the performance of the feature a little bit, it would also likely produce a less maintainable code pattern. This code is in the hot-path so performance is definitely important to keep front and center.... 🤔 

Having the source type in `Packet` helps us make decisions on behavior if we need to with helper functions, as opposed to the other approach where we add (ultimately an arbitrary number of) boolean fields to the struct to guide the logic. 

If performance impact is negligible, I think I prefer the **current** implementation.  The go compiler might even inline the `eolEnabled()` function (I'm not sure - might be worth checking), making the performance overhead a non-issue. 

I'm good with this. Nice job.

File: pkg/dogstatsd/listeners/packet_assembler.go [hunk 2]
```diff
@@ -22,10 +22,11 @@ type packetAssembler struct {
 	sharedPacketPool *PacketPool
 	flushTimer       *time.Ticker
 	closeChannel     chan struct{}
+	packetSourceType SourceType
 	sync.Mutex
 }
 
-func newPacketAssembler(flushTimer time.Duration, packetsBuffer *packetsBuffer, sharedPacketPool *PacketPool) *packetAssembler {
+func newPacketAssembler(flushTimer time.Duration, packetsBuffer *packetsBuffer, sharedPacketPool *PacketPool, packetSourceType SourceType) *packetAssembler {
```
Comment (ogaca-dd) [inline pkg/dogstatsd/listeners/packet_assembler.go:29]: `packetSourceType` is not used.

**Comment by ogaca-dd** [comment]
@truthbk 

> I am thinking about @ogaca-dd comment.... while his comment might help improve the performance of the feature a little bit, it would also likely produce a less maintainable code pattern. This code is in the hot-path so performance is definitely important to keep front and center.... 🤔

I have pushed a quick implementation of what I suggested. https://github.com/DataDog/datadog-agent/commit/8eb17f4a06bd687c0287aad4a82bbe4978e54ee4 (It is not tested and it must not be cherry-picked as if).
IMO, the code is simpler.
Note: Unit tests may require several changes because of a lack of DRY.

> Having the source type in Packet helps us make decisions on behavior if we need to with helper functions, as opposed to the other approach where we add (ultimately an arbitrary number of) boolean fields to the struct to guide the logic.

I do not think handling the EOL is something fundamentally related to the source type. In the current code, it is related to the source type because the config option is defined by source type. That explain why in the previous implementation, there is no need of the source type. IMO, `Packet` should not have the source type to handle the EOL. 

The current implementation is fine, I wanted to explain why I suggested to remove the source type from `Packet`.

**Comment by gh123man** [comment]
Thank you both for the feedback. 
My intention behind the current implementation was that this enables us to make decisions on the packet based on the source type anywhere in the pipeline. This could potentially be used for future features (if any), and also keeps the EOL handling code all in one place. Also since the listeners are defined in the same package as `types` it seems reasonable for `Packet` to be made aware of this context. 

I do recognize that passing the EOL boolean though the packet will end up being less code though :)

**Review by truthbk — APPROVED** [review]
Looks good to me! 👍 

Not sure why the CI failed, probably worth checking before merging.

**Review by ogaca-dd — APPROVED** [review]

**Pull request merged**